### PR TITLE
executor: fix insert on duplicate update

### PR DESCRIPTION
**Pull request by jackysp**
Do insert when check duplicate and update. It is not the same logic as `insert ignore`, because `insert ignore` only skip the duplicate rows, without doing anything.
PTAL @coocood

### CODE REVIEW

**Comment by shenli** [comment]
@winkyao @zhexuany @coocood PTAL

**Comment by coocood** [comment]
LGTM

File: executor/write.go [hunk 1]
```diff
@@ -817,59 +834,60 @@ func (e *InsertExec) exec(ctx context.Context, rows [][]types.Datum) (Row, error
...
-			if len(e.OnDuplicate) > 0 && e.IgnoreErr {
-				data, err1 := e.Table.RowWithCols(e.ctx, h, e.Table.WritableCols())
-				if err1 != nil {
-					return nil, errors.Trace(err1)
-				}
-				if _, _, _, err = e.doDupRowUpdate(h, data, row, e.OnDuplicate); err != nil {
-					return nil, errors.Trace(err)
+		for _, row := range rows {
+			// duplicate row will be marked as nil in batchMarkDupRows if
+			// IgnoreErr is true. For IgnoreErr is false, it is a protection.
```
Comment (zhexuany) [inline executor/write.go]: “For IgnoreErr being false, it is a protection." ? Please also specify what is this protection.

**Comment by shenli** [comment]
@jackysp Please address the comment.

**Comment by shenli** [comment]
Please resolve the conflicts.

File: executor/write.go (hunk 1 above)
Comment (jackysp) [reply executor/write.go]: `batchMarkDupRows` will set row = nil to mark the row is duplicated, so we need to skip the nil row here. If ignore error is false, the row should not be nil, but we still check  
 nil here, so it is the protection.

**Pull request merged**